src/system/SystemObject.h

```diff
@@ -213,7 +231,13 @@ inline T * ObjectPool<T, N>::Get(const Layer & aLayer, size_t aIndex)
     T * lReturn = nullptr;
 
     if (aIndex < N)
```

Comment:
Should this condition be checked for the heap implementation?

Reply:
Yes, we do not have this limitation with heap implementation, but I still treat N as the maximum capacity in heap implementation. The idea is the underneath implementation with heap or static allocation should be invisible to upper layer, syntax of interface should be same.

If we remove N check for heap implementation, we need to rewrite some test cases just for heap implementation since we never hit pool full case. Some upper layer logic may also need to be adjusted accordingly.

Reply:
The whole point is to allow some version of "unlimited" for the heap implementation, right?  Either by default or at least optionally.  I guess you can sort of get there with `N == static_cast<unsigned int>(-1)` but it's a little non-obvious.

Related question: do we want this to be an all-or-nothing switch like in this PR, where all pools get switched with a single ifdef, or would it make more sense to have a magic value (0? -1?) for the pool size that switches over to a heap implementation and then consumers can decide on a per-pool basis whether to just use heap (and unlimited sizing) or a fixed-size pool?

And do we want to support any sort of "preallocate N entries but then allow spilling over into the heap" kind of setup?  That seems kind of complicated, so my temptation would be "no".

Reply:
If we don't want to check against N, then we don't need to do the check and rely on the success/fail of allocation.

Whether to use heap or fixed-size pool is depending on the platform capability, right? heap based implementation is only viable on large system such as Linux. I think this should be statically configured.

Reply:
I don't agree.  If you set N to 10,000,000, this doesn't mean that allocation will succeed.  That's a problem either way.  So I don't really see the value of capping at N.  It just forces implementations to try to predict ahead of time what memory will be needed and what will be available, which is the whole problem really.

I think if you're ever calling 'new' in your program, you must simply be comfortable with the fact that allocation may fail and constructors will throw exceptions.  Then you have to decide whether you're going to catch these.

So I'd say that's our choice.  I personally think the N-check can go away and our only question here is whether we should try to catch exceptoiins.

Reply:
Removed the ceiling check for heap based implementation